src/python/middle_step.py

```diff
@@ -0,0 +1,67 @@
+from z3 import Implies, And, Or  # type: ignore
+
+from typing import Any, List
+
+
+def binary_comb_z3(list1: List[Any], list2: List[Any]) -> List[Any]:
```

Comment:
Can we remove any and be more explicit?

Reply:
honestly, I haven't found what type those things are